code/modules/pulsar_engine/pulsar_consoles.dm

```diff
@@ -0,0 +1,269 @@
+	data["power_produced"] = get_produced_power()
+	data["ship_fuel"] = round(tank?.air_contents.get_total_moles())
+	data["no_tank"] = !tank
+	data["thrust_cost"] = get_thrust_cost()
+	data["shield_power"] = shield_power
+	data["shield_power_req"] = get_required_shielding()
+	data["effective_power_produced"] = get_effective_power_porduced()
+
+	return data
+
+/obj/machinery/pulsar/Topic(href, href_list)
+	if(href_list["move"])
+		if(tank)
+			var/thrust_cost = get_thrust_cost()
+			if(tank.air_contents.get_total_moles() >= thrust_cost)
+				tank.air_contents.remove(thrust_cost)
+				var/newdir = text2num(href_list["move"])
+				ship.try_move(newdir, TRUE)
+
+	else if(href_list["scan_fuel"])
+		scan_for_fuel()
+	
+	else if(href_list["set_shield"])
+		var/target_level = input(usr, "Set shielding power", "Shield control", 50) as num
+		if(target_level < 100)
+			shield_power = max(0 , target_level)
+
+	SSnano.update_uis(src)
+	..()
+
+/obj/machinery/pulsar/proc/get_thrust_cost()
+	if(tank?.air_contents.specific_mass())
+		return round((1/tank.air_contents.specific_mass()) * 2) //Goes from 4 with 100% plasma up to 40+ when diluting with too much oxygen
+	return 100
+
+/obj/machinery/pulsar/proc/get_produced_power() //Simply based on linear distance from the pulsar
+	for(var/obj/O in get_turf(ship))
+		if(O.type in subtypesof(/obj/effect/pulsar_beam))
+			return 150
+	return max(0, round(((GLOB.maps_data.pulsar_size * ROOT(2,2)) - 2 * ROOT(2, abs(linked.x - ship.x) ** 2 + abs(linked.y - ship.y) ** 2)) * 100/(GLOB.maps_data.pulsar_size * ROOT(2,2))))
+
+/obj/machinery/pulsar/proc/get_required_shielding()
+	for(var/obj/O in get_turf(ship))
+		if(O.type in subtypesof(/obj/effect/pulsar_beam))
+			return 60
+	return max(30, 30 + (round(((GLOB.maps_data.pulsar_size * ROOT(2,2)) - 2 * ROOT(2, abs(linked.x - ship.x) ** 2 + abs(linked.y - ship.y) ** 2)) * 100/(GLOB.maps_data.pulsar_size * ROOT(2,2))) / 5))
+
+/obj/machinery/pulsar/proc/get_effective_power_porduced() //Formula comes with deminishing returns but it's rising way past 100
+	return get_produced_power() * (100 - shield_power) / 100
+
+/obj/machinery/pulsar/proc/onShipMoved()
+	SSnano.update_uis(src)
+
+/obj/machinery/pulsar/proc/scan_for_fuel()
+	tank = locate(/obj/structure/pulsar_fuel_tank) in get_area(src)
+	tank?.connected_console = src
+	SSnano.update_uis(src)
+
+/obj/machinery/power/pulsar_power_bridge //Only holds ref to the console and its area, used to get power from it, or disconnect the ship.
+	name = "pulsar power bridge"
+	icon = 'icons/obj/computer.dmi'
+	icon_state = "computer"
+	density = TRUE
+	anchored = TRUE
+	layer = BELOW_OBJ_LAYER
+	var/obj/machinery/pulsar/pulsar_console
+	var/area/console_area
+
+/obj/machinery/power/pulsar_power_bridge/Initialize(mapload, d)
+	connect_to_network()
+	return INITIALIZE_HINT_LATELOAD
+
+/obj/machinery/power/pulsar_power_bridge/LateInitialize()
+	. = ..()
+	pulsar_console = locate() in world //I can get away with it once, right?
```

Comment:
Consider instead assigning a tag to the console.

Reply:
https://www.youtube.com/watch?v=L6z6A5TP7uA